### PR TITLE
fix: do not draw the buffer object if glBufferData failed

### PR DESCRIPTION
**Pull request by wzpan**
If `glBufferData()` failed, it's easily to get SIGSEGV crash when calling `glDrawElements()`.

This PR add a `glGetError()` checking after calling `glBufferData()` to avoid calling `glDrawElements()` after `glBufferData()` failed.

Related discussion: https://forum.cocos.org/t/topic/104524/6

### CODE REVIEW

File: cocos/renderer/gfx/IndexBuffer.cpp [hunk 1]
```diff
@@ -110,11 +111,25 @@ void IndexBuffer::update(uint32_t offset, const void* data, size_t dataByteLengt
...
     else
     {
         glBufferSubData(GL_ELEMENT_ARRAY_BUFFER, (GLintptr)offset, (GLsizeiptr)dataByteLength, (const GLvoid*)data);
+        GLenum gl_err = glGetError();
+        if (0 != gl_err){
+            RENDERER_LOGE("GL error 0x%x: %s:%s", gl_err, glEnumName(gl_err), __FUNCTION__);
+            _drawable = false;
+        }else{
+            _drawable = true;
+        }
```
Comment (minggo) [inline cocos/renderer/gfx/IndexBuffer.cpp:132]: Duplicate codes. I know may want to log the function line, but can use `_needExpandDataStore ` to know which codes are executed.

File: cocos/renderer/gfx/IndexBuffer.cpp [hunk 2]
```diff
@@ -110,11 +111,25 @@ void IndexBuffer::update(uint32_t offset, const void* data, size_t dataByteLengt
...
         _needExpandDataStore = false;
     }
     else
     {
         glBufferSubData(GL_ELEMENT_ARRAY_BUFFER, (GLintptr)offset, (GLsizeiptr)dataByteLength, (const GLvoid*)data);
+        GLenum gl_err = glGetError();
```
Comment (minggo) [inline cocos/renderer/gfx/IndexBuffer.cpp:126]: Please use glError instead to obey the coding style.

File: cocos/renderer/gfx/DeviceGraphics.cpp [hunk 3]
```diff
@@ -409,6 +409,15 @@ void DeviceGraphics::draw(size_t base, GLsizei count)
     }
     
     commitTextures();
+
+    if(!nextIndexBuffer->isDrawable()){
+        _drawCalls++;
+        auto temp = _nextState;
+        _nextState = _currentState;
+        _currentState = temp;
```
Comment (minggo) [inline cocos/renderer/gfx/DeviceGraphics.cpp:417]: use std::swap() is better

File: cocos/renderer/gfx/DeviceGraphics.cpp [hunk 4]
```diff
@@ -409,6 +409,15 @@ void DeviceGraphics::draw(size_t base, GLsizei count)
     }
     
     commitTextures();
+
+    if(!nextIndexBuffer->isDrawable()){
```
Comment (minggo) [inline cocos/renderer/gfx/DeviceGraphics.cpp:413]: May be can check it as early as possible. For example, add after `auto nextIndexBuffer = _nextState->getIndexBuffer();`.

File: cocos/renderer/gfx/IndexBuffer.cpp [hunk 5]
```diff
@@ -110,11 +111,25 @@ void IndexBuffer::update(uint32_t offset, const void* data, size_t dataByteLengt
     {
 //        glBufferData(GL_ELEMENT_ARRAY_BUFFER, (GLsizeiptr)dataByteLength, data, glUsage);
         glBufferData(GL_ELEMENT_ARRAY_BUFFER, _bytes, (const GLvoid*)data, glUsage);
+        GLenum gl_err = glGetError();
+        if (0 != gl_err){
+            RENDERER_LOGE("GL error 0x%x: %s:%s", gl_err, glEnumName(gl_err), __FUNCTION__);
+            _drawable = false;
```
Comment (minggo) [inline cocos/renderer/gfx/IndexBuffer.cpp:117]: Should initialize it in `init()`.

Comment (minggo) [reply cocos/renderer/gfx/IndexBuffer.cpp:117]: As the object will be destroyed and initialized again.

File: cocos/renderer/gfx/IndexBuffer.cpp (hunk 1 above)
Comment (wzpan) [reply cocos/renderer/gfx/IndexBuffer.cpp:132]: Actually both blocks will set `_needExpandDataStore` as false after execution. 
But it's fine to catch them all in one piece of code. I will modify it a little bit.

File: cocos/renderer/gfx/IndexBuffer.cpp (hunk 5 above)
Comment (wzpan) [reply cocos/renderer/gfx/IndexBuffer.cpp:117]: You meant init `gl_err` object? I can't find any example code doing that. How?

Comment (minggo) [reply cocos/renderer/gfx/IndexBuffer.cpp:117]: Sorry, i mean `_drawable`.

**Comment by minggo** [comment]
I think it is better to check gl error in debug mode, as it is time consuming.

File: cocos/renderer/gfx/IndexBuffer.cpp (hunk 5 above)
Comment (wzpan) [reply cocos/renderer/gfx/IndexBuffer.cpp:117]: I see. Gotcha.

**Comment by wzpan** [comment]
> I think it is better to check gl error in debug mode, as it is time consuming.

But if so how can we catch the crash in release version?

**Comment by minggo** [comment]
> > I think it is better to check gl error in debug mode, as it is time consuming.
> 
> But if so how can we catch the crash in release version?

The invoker should check the data before updating.

**Comment by minggo** [comment]
> If glBufferData() failed, it's easily to get SIGSEGV crash when calling glDrawElements().

What kind of error of glBufferData() will cause SIGSEGV?

**Comment by wzpan** [comment]
> > If glBufferData() failed, it's easily to get SIGSEGV crash when calling glDrawElements().
> 
> What kind of error of glBufferData() will cause SIGSEGV?

As you can see in the topic: https://forum.cocos.org/t/topic/104524/6

`glBufferData()` itself won't cause SIGSEGV, but `glDrawElements()` do when trying to draw bad buffer objects. We need a way to spot the case.

**Comment by minggo** [comment]
> but glDrawElements() do when trying to draw bad buffer objects

I can not find crash information in [the doc](https://docs.gl/es2/glDrawElements). Could you please provide an example to reproduce it?

**Comment by wzpan** [comment]
> > but glDrawElements() do when trying to draw bad buffer objects
> 
> I can not find crash information in [the doc](https://docs.gl/es2/glDrawElements). Could you please provide an example to reproduce it?

Sadly the crash happens occasionally and we also cannot find a reproduceable example. However if you revert the PR #2884 and try the demo of #2930 in iOS you will get a similar crash.

**Comment by wzpan** [comment]
@minggo I've submitted the change according to the review option. But I have to say that I do this at my home since it's holiday. Hence I haven't fully test it. Please test them further if you decide to merge it.

**Comment by minggo** [comment]
> > > but glDrawElements() do when trying to draw bad buffer objects
> > 
> > 
> > I can not find crash information in [the doc](https://docs.gl/es2/glDrawElements). Could you please provide an example to reproduce it?
> 
> Sadly the crash happens occasionally and we either cannot find a reproduceable example. However if you revert the PR #2884 and try the demo of #2930 in iOS you will get a similar crash.

I can understand the modification of #2884, as passed in data is invalid. Except that, i do not know why it will cause crash in draw elements. If passed in pointer doesn't contain enough data, then i think we should check the passed data in binding codes.

**Comment by minggo** [comment]
> @minggo I've submitted the change according to the review option. But I have to say that I do this at my home since it's holiday. Hence I haven't fully test it. Please test them further if you decide to merge it.

Thanks.

**Comment by wzpan** [comment]
> > > > but glDrawElements() do when trying to draw bad buffer objects
> > > 
> > > 
> > > I can not find crash information in [the doc](https://docs.gl/es2/glDrawElements). Could you please provide an example to reproduce it?
> > 
> > 
> > Sadly the crash happens occasionally and we either cannot find a reproduceable example. However if you revert the PR #2884 and try the demo of #2930 in iOS you will get a similar crash.
> 
> I can understand the modification of #2884, as passed in data is invalid. Except that, i do not know why it will cause crash in draw elements. If passed in pointer doesn't contain enough data, then i think we should check the passed data in binding codes.

I agree. Do the checking of passed data is prefered for performance concern.

**Pull request closed without merging**